src/audio/component.c

```diff
@@ -328,162 +328,286 @@ int comp_verify_params(struct comp_dev *dev, uint32_t flag,
-	 */
-	model->data_pos += cdata->num_elems;
+		/* If component state is READY we can omit old
+		 * configuration immediately. When in playback/capture
+		 * the new configuration presence is checked in copy().
+		 */
+		if (dev->state ==  COMP_STATE_READY) {
+			rfree(blob_handler->data);
+			blob_handler->data = NULL;
+		}
 
-	/* Update crc value when done */
-	if (done) {
-		model->crc = crc32(0, model->data, model->data_size);
-		comp_dbg(dev, "comp_set_model() done, memory_size = 0x%x, crc = 0x%08x",
-			 model->data_size, model->crc);
+		/* If there is no existing configuration the received
+		 * can be set to current immediately. It will be
+		 * applied in prepare() when streaming starts.
+		 */
+		if (!blob_handler->data) {
+			blob_handler->data = blob_handler->data_new;
+			blob_handler->data_new = NULL;
+		}
 	}
 
 	return 0;
 }
 
-int comp_get_model(struct comp_dev *dev, struct comp_model_data *model,
-		   struct sof_ipc_ctrl_data *cdata, int size)
+int comp_data_blob_get_cmd(struct comp_dev *dev,
+			   struct comp_data_blob_handler *blob_handler,
+			   struct sof_ipc_ctrl_data *cdata, int size)
 {
-	size_t bs;
 	int ret = 0;
 
-	comp_dbg(dev, "comp_get_model() msg_index = %d, num_elems = %d, remaining = %d ",
+	comp_dbg(dev, "comp_data_blob_get_cmd() msg_index = %d, num_elems = %d, remaining = %d ",
 		 cdata->msg_index, cdata->num_elems,
 		 cdata->elems_remaining);
 
 	/* Copy back to user space */
-	if (model->data) {
+	if (blob_handler->data) {
 		/* reset data_pos variable in case of copying first element */
 		if (!cdata->msg_index) {
-			model->data_pos = 0;
-			comp_dbg(dev, "comp_get_model() model data_size = 0x%x",
-				 model->data_size);
+			blob_handler->data_pos = 0;
+			comp_dbg(dev, "comp_data_blob_get_cmd() model data_size = 0x%x",
+				 blob_handler->data_size);
 		}
 
-		bs = cdata->num_elems;
-
 		/* return an error in case of mismatch between num_elems and
 		 * required size
 		 */
-		if (bs > size) {
-			comp_err(dev, "comp_get_model(): invalid size %d", bs);
+		if (cdata->num_elems > size) {
+			comp_err(dev, "comp_data_blob_get_cmd(): invalid cdata->num_elems %d",
+				 cdata->num_elems);
 			return -EINVAL;
 		}
 
 		/* copy required size of data */
 		ret = memcpy_s(cdata->data->data, size,
-			       (char *)model->data + model->data_pos,
-			       bs);
+			       (char *)blob_handler->data + blob_handler->data_pos,
+			       cdata->num_elems);
 		assert(!ret);
 
 		cdata->data->abi = SOF_ABI_VERSION;
-		cdata->data->size = model->data_size;
-		model->data_pos += bs;
-
+		cdata->data->size = blob_handler->data_size;
+		blob_handler->data_pos += cdata->num_elems;
 	} else {
-		comp_warn(dev, "comp_get_model(): model->data not allocated yet.");
+		comp_warn(dev, "comp_data_blob_get_cmd(): model->data not allocated yet.");
 		cdata->data->abi = SOF_ABI_VERSION;
 		cdata->data->size = 0;
 	}
 
 	return ret;
 }
 
+struct comp_data_blob_handler *comp_data_blob_handler_new(struct comp_dev *dev)
```

Comment:
This function just calls alloc() and struct size is known to the caller. Not sure if we want to keep this code just in case another step to be added in future.

Reply:
@mmaka1 
In that case struct size is unknown to caller. It is declared in .c file and in .h file we have only forward declaration.

Reply:
Right, full data encapsulation is applied here and 'constructor' is required. Nicely done.